pkg-config.cmake

```diff
@@ -220,17 +220,30 @@ MACRO(_SETUP_PROJECT_PKG_CONFIG_FINALIZE)
   ENDIF()
 ENDMACRO(_SETUP_PROJECT_PKG_CONFIG_FINALIZE)
 
-# _PARSE_PKG_CONFIG_STRING (PKG_CONFIG_STRING _PKG_LIB_NAME_VAR _PKG_PREFIX_VAR)
+# _PARSE_PKG_CONFIG_STRING (PKG_CONFIG_STRING _PKG_LIB_NAME_VAR _PKG_PREFIX_VAR _PKG_CONFIG_STRING_NOSPACE_VAR)
 # ----------------------------------------------------------
 # 
-# Retrieve the library name and the prefix used for CMake variable names
-# from the pkg-config string. For instance, `my-package > 0.4` results in
+# Retrieve from the pkg-config string:
+# - the library name,
+# - the prefix used for CMake variable names,
+# - a variant without spaces around the operator (if there is an operator),
+#   as expected by cmake's CHECK_PKG_MODULE.
+# . For instance, `my-package > 0.4` results in
 # - _PKG_LIB_NAME_VAR <- my-package
 # - _PKG_PREFIX_VAR <- MY_PACKAGE
-MACRO(_PARSE_PKG_CONFIG_STRING PKG_CONFIG_STRING _PKG_LIB_NAME_VAR _PKG_PREFIX_VAR)
-  # Retrieve the left part of the equation to get package name.
-  STRING(REGEX MATCH "[^<>= ]+" ${_PKG_LIB_NAME_VAR} "${PKG_CONFIG_STRING}")
-  # And transform it into a valid variable prefix.
+# - _PKG_CONFIG_STRING_NOSPACE_VAR <- `my-package>0.4`
+# `my-package` results in
+# - _PKG_LIB_NAME_VAR <- my-package
+# - _PKG_PREFIX_VAR <- MY_PACKAGE
+# - _PKG_CONFIG_STRING_NOSPACE_VAR <- `my-package`
+MACRO(_PARSE_PKG_CONFIG_STRING PKG_CONFIG_STRING _PKG_LIB_NAME_VAR _PKG_PREFIX_VAR _PKG_CONFIG_NOSPACE_VAR)
+  # Decompose the equation
+  STRING(REGEX MATCH "([^ ]+)( (>|>=|=|<=|<) (.*))?" _FOO "${PKG_CONFIG_STRING}")
```

Comment:
Shouldn't it be `([^ ]+)( ?(>|>=|=|<=|<) ?(.*))?` ? So that it supports the case with no spaces.

Moreover, could you rename `_FOO` to `_UNUSED` ?

Reply:
> Shouldn't it be `([^ ]+)( ?(>|>=|=|<=|<) ?(.*))?` ? So that it supports the case with no spaces.

cfr above, my plan is not to add support for the case without spaces.

Is it five with you?

> Moreover, could you rename `_FOO` to `_UNUSED` ?
done.

Thanks a lot for the review!

Reply:
> cfr above, my plan is not to add support for the case without spaces.

It might be both surprising and annoying for projects that does `ADD_REQUIRED_DEPENDENCY("foo>1"))`. Do you know what error message they will get ?

If it is something like "*package 'foo>1' was not found*" then it is really tricky because it is fairly hard to understand that the issue is the missing space...

Reply:
> > cfr above, my plan is not to add support for the case without spaces.
> 
> It might be both surprising and annoying for projects that does `ADD_REQUIRED_DEPENDENCY("foo>1"))`. Do you know what error message they will get ?

without this pull request, the project would build with success, but generate an invalid ".pc".
dependers would fail at cmake run with

```
CMake Error at /usr/share/cmake-3.10/Modules/FindPkgConfig.cmake:415 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.10/Modules/FindPkgConfig.cmake:593 (_pkg_check_modules_internal)
  cmake/pkg-config.cmake:298 (PKG_CHECK_MODULES)
  cmake/pkg-config.cmake:510 (ADD_DEPENDENCY)
  CMakeLists.txt:36 (ADD_REQUIRED_DEPENDENCY)
```
(this just happened to pinocchio dependers who use devel, see
https://github.com/stack-of-tasks/pinocchio/pull/933#issuecomment-551333575)

I can try to build example-robot-data next week and see if this commit would change anything with this respect.

> If it is something like "_package 'foo>1' was not found_" then it is really tricky because it is fairly hard to understand that the issue is the missing space...

I agree, I don't think this pull request makes the matter worse, but I can check next week.

By the way, PKG_CHECK_MODULES' implementation seems incomplete. If I read it correctly, it would parse "toto>=3" and convert it into "toto >=3", but not "toto>3". (I don't think we need to do anything about it)

Reply:
> without this pull request, the project would build with success, but generate an invalid ".pc".

Thanks for the details. I don't think you have to worry more than this.